core/src/main/java/org/apache/druid/java/util/emitter/core/HttpPostEmitter.java

```diff
@@ -698,13 +698,42 @@ public boolean apply(Throwable e)
       catch (Exception e) {
         if (e == timeoutLessThanMinimumException) {
           log.debug(e, "Failed to send events to url[%s] with timeout less than minimum", config.getRecipientBaseUrl());
+          sendBackoffDelay();
         } else {
           log.error(e, "Failed to send events to url[%s]", config.getRecipientBaseUrl());
         }
         return false;
       }
     }
 
+    /**
+     * Back-off the sender for a period of time.
+     *
+     * Call only from the send() loop thread when it sees a timeout.
+     */
+    void sendBackoffDelay()
+    {
+      // The backoff delay is quite simple in favour of something more complicated.
+
+      // This is called when the send() call hits a timeoutLessThanMinimumException
+      // condition. Ideally, what we should do is delay until either the timeout increases
+      // to above the minHttpTimeout, and so the send() will be issued, or, until a Batch
+      // wants to flush based on the configured timeout.
+
+      // Making this technically correct, across all possible Batches, including large events,
+      // etc is complicated in this codebase, since batch flush timing is not tracked here,
+      // so rather than introduce new bugs, let's keep it simple.
+
+      final long backoffCheckDelayMillis = config.getMinHttpTimeoutMillis() / 5;
```

Comment:
I'm still confused. Why not keep the computed timeout in the exception object and wait for (config.getMinHttpTimeoutMillis - computed_timeout)?

Reply:
It's a heuristic, I don't have a good reason for `/ 5` instead of 4, 3, etc.

WRT sleep timer - it's definitely possible to sleep for longer until the timeout expires, but the downside is we might wait a little longer than required for an opportunity to send if the computed timeout changes. For example, if the batch fills while we are still sleeping then we might keep sleeping. We could resolve this by introducing an interrupt or a wait/notify, etc.

WRT passing the time calculated, it adds more code to maintain a new custom Exception for (in my view) limited benefit. For what I assume are efficiency reasons the existing code [caches the Exception object here](https://github.com/apache/druid/blob/5ab9ec6bdcc06078d678b88903b4d53bff2c9d61/core/src/main/java/org/apache/druid/java/util/emitter/core/HttpPostEmitter.java#L478). To be honest I don't really see the benefit of caching the object, and would suggest an easier solution would be to recalculate the remaining timeout directly.

--

The main goal when I started was to avoid busy-spinning an entire CPU core as it was previously doing, a sleep allows this while still being response enough, and this is the minimum viable solution I came up with. 

IF we do a more comprehensive solution solution I think we completely split the logic between batch accumulation and flush / send pacing / backpressure. Also, that's substantially more code for what is limited real-world benefit.

Reply:
Hi @FrankChen021 were you able to review / confirm suitability of the approach here?

CC @abhishekagarwal87

Reply:
Let me review the changes on this weekend. Thanks for the patience.